flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala

```diff
@@ -20,90 +20,180 @@ package org.apache.flink.table.api
+    *
+    * @param preceding preceding offset relative to the current row.
+    * @return an over window with defined preceding
+    */
+  def preceding(preceding: Expression): OverWindowPartitionedOrderedPreceding = {
+    new OverWindowPartitionedOrderedPreceding(partitionBy, orderBy, preceding)
+  }
+
+  /**
+    * Assigns an alias for this window that the following `select()` clause can refer to.
+    *
+    * @param alias alias for this over window
+    * @return the fully defined over window
+    */
+  def as(alias: String): OverWindow = as(ExpressionParser.parseExpression(alias))
+
+  /**
+    * Assigns an alias for this window that the following `select()` clause can refer to.
+    *
+    * @param alias alias for this over window
+    * @return the fully defined over window
+    */
+  def as(alias: Expression): OverWindow = {
+    new OverWindow(alias, partitionBy, orderBy, UnboundedRange(), None)
+  }
+}
 
 /**
-  * A partially defined over window.
+  * Partially defined over window with (optional) partitioning, order, and preceding.
   */
-class OverWindowWithPreceding(
+class OverWindowPartitionedOrderedPreceding(
     private val partitionBy: Seq[Expression],
     private val orderBy: Expression,
     private val preceding: Expression) {
 
-  private[flink] var following: Expression = _
+  private var optionalFollowing: Option[Expression] = None
 
   /**
     * Assigns an alias for this window that the following `select()` clause can refer to.
     *
     * @param alias alias for this over window
-    * @return over window
+    * @return the fully defined over window
     */
   def as(alias: String): OverWindow = as(ExpressionParser.parseExpression(alias))
 
   /**
     * Assigns an alias for this window that the following `select()` clause can refer to.
     *
     * @param alias alias for this over window
-    * @return over window
+    * @return the fully defined over window
     */
   def as(alias: Expression): OverWindow = {
-
-    // set following to CURRENT_ROW / CURRENT_RANGE if not defined
-    if (null == following) {
-      if (preceding.resultType.isInstanceOf[RowIntervalTypeInfo]) {
-        following = CURRENT_ROW
-      } else {
-        following = CURRENT_RANGE
-      }
-    }
-    OverWindow(alias, partitionBy, orderBy, preceding, following)
+    new OverWindow(alias, partitionBy, orderBy, preceding, optionalFollowing)
   }
 
   /**
     * Set the following offset (based on time or row-count intervals) for over window.
     *
-    * @param following following offset that relative to the current row.
-    * @return this over window
+    * @param followingExpr following offset that relative to the current row.
+    * @return an over window with defined following
     */
-  def following(following: String): OverWindowWithPreceding = {
-    this.following(ExpressionParser.parseExpression(following))
+  def following(followingExpr: String): OverWindowPartitionedOrderedPreceding = {
```

Comment:
Should we rename these xxxExpr to xxx? It seems strange to add the "Expr" suffix for the String parameters.

Reply:
+1

Reply:
+1, my IDE was just complaining about the name duplicate but I solved this issue by using `this.` again.